bpf/kmesh/workload/include/authz.h

```diff
@@ -32,79 +49,187 @@ static inline wl_policies_v *get_workload_policies_by_uid(__u32 workload_uid)
+        if ((void *)(info->ip6h + 1) > end || (info->ip6h->nexthdr != IPPROTO_TCP))
+            return PARSER_FAILED;
+        begin = (info->ip6h + 1);
+    } else
+        return PARSER_FAILED;
+
+    info->tcph = (struct tcphdr *)begin;
+    if ((void *)(info->tcph + 1) > end)
+        return PARSER_FAILED;
+    return PARSER_SUCC;
+}
+
+static inline void parser_tuple(struct xdp_info *info, struct bpf_sock_tuple *tuple_info)
 {
+    if (info->iph->version == 4) {
+        tuple_info->ipv4.saddr = info->iph->saddr;
+        tuple_info->ipv4.daddr = info->iph->daddr;
+        tuple_info->ipv4.sport = info->tcph->source;
+        tuple_info->ipv4.dport = info->tcph->dest;
+    } else {
+        bpf_memcpy((__u8 *)tuple_info->ipv6.saddr, info->ip6h->saddr.in6_u.u6_addr8, IPV6_ADDR_LEN);
+        bpf_memcpy((__u8 *)tuple_info->ipv6.daddr, info->ip6h->daddr.in6_u.u6_addr8, IPV6_ADDR_LEN);
+        tuple_info->ipv6.sport = info->tcph->source;
+        tuple_info->ipv6.dport = info->tcph->dest;
+    }
+}
+
+static inline int get_tuple_key(struct xdp_md *ctx, struct bpf_sock_tuple *tuple_info, struct xdp_info *info)
+{
+    int ret = parser_xdp_info(ctx, info);
+    if (ret != PARSER_SUCC) {
+        BPF_LOG(ERR, AUTH, "Failed to parse xdp_info\n");
+        return PARSER_FAILED;
+    }
+
+    parser_tuple(info, tuple_info);
+
+    return PARSER_SUCC;
+}
+
+SEC("xdp_auth")
+int matchDstPorts(struct xdp_md *ctx)
+{
+    struct match_ctx *res;
     __u32 *notPorts = NULL;
     __u32 *ports = NULL;
     __u32 i;
+    __u16 dport = 0;
+    Istio__Security__Match *match = NULL;
+    struct bpf_sock_tuple tuple_key = {0};
+    struct xdp_info info = {0};
+    int ret;
+
+    if (get_tuple_key(ctx, &tuple_key, &info) != PARSER_SUCC) {
+        BPF_LOG(ERR, AUTH, "Failed to get tuple key\n");
+        return XDP_ABORTED;
+    }
+
+    res = bpf_map_lookup_elem(&tailcall_info_map, &tuple_key);
+    if (!res) {
+        BPF_LOG(ERR, AUTH, "Failed to retrieve res from map\n");
+        return XDP_PASS;
+    }
+
+    if (info.iph->version == 4) {
+        dport = tuple_key.ipv4.dport;
+    } else if (info.ip6h->version == 6) {
+        dport = tuple_key.ipv6.dport;
+    } else {
+        BPF_LOG(ERR, AUTH, "Invalid IP version.\n");
+        return XDP_PASS;
+    }
+    match = (Istio__Security__Match *)kmesh_get_ptr_val(res->match);
+    if (!match) {
+        BPF_LOG(ERR, AUTH, "match pointer is null\n");
+        return XDP_PASS;
+    }
 
     if (match->n_destination_ports == 0 && match->n_not_destination_ports == 0) {
-        return MATCHED;
+        bpf_tail_call(ctx, &xdp_tailcall_map, TAIL_CALL_AUTH_IN_USER_SPACE);
+        return (res->action == AUTH_DENY) ? XDP_PASS : XDP_DROP;
     }
 
     if (match->n_not_destination_ports != 0) {
         notPorts = kmesh_get_ptr_val(match->not_destination_ports);
         if (!notPorts) {
-            return UNMATCHED;
+            return (res->action == AUTH_DENY) ? XDP_PASS : XDP_DROP;
```

Comment:
If the configuration is abnormal, it is more reasonable to directly bypass the service.